ribbon/src/test/java/feign/ribbon/RibbonClientTest.java

```diff
@@ -51,6 +57,26 @@
   @Rule
   public final MockWebServer server2 = new MockWebServer();
 
+  private static String oldRetryConfig = null;
+
+  private static final String SUN_RETRY_PROPERTY = "sun.net.http.retryPost";
+
+  @BeforeClass
+  public static void disableSunRetry() throws Exception {
+    // The Sun HTTP Client retries all requests once on an IOException, which makes testing retry code harder than would
+    // be ideal. We can only disable it for post, so lets at least do that.
+    oldRetryConfig = System.setProperty(SUN_RETRY_PROPERTY, "false");
```

Comment:
Neither did I until I wrote that test.

It's all about handling Keep-Alive connections. It'd be unfair on the application to propagate the IOException if it's caused by the reuse of a Keep-Alive connection that happened to be shutdown by the server (or a NAT/firewall timeout) at exactly the wrong moment, so the client will retry once when it looks like that might have happened (i.e. an IO error while writing a request or reading the response headers.)

This behaviour appears to be explicitly permitted, even for methods that aren't idempotent, by [RFC7230 section 6.3.2](https://tools.ietf.org/html/rfc7230#section-6.3.2).
Out of curiosity I had a quick dig through the Chromium source, and, assuming I'm reading it correctly, it too follows this behaviour. The Apache commons client, meanwhile, I think wont for a POST by default, which is why it'll sometimes throw inexplicable connection errors if keepalives are in use.